STL_Extension/include/CGAL/iterator.h

```diff
@@ -37,28 +37,34 @@
 
 namespace CGAL {
 
-template<typename I>
+template<typename I, typename Value_type = I>
 class Prevent_deref
   : public boost::iterator_adaptor<
-  Prevent_deref<I>
+    Prevent_deref<I,Value_type>
   , I // base
-  , I // value
+  , Value_type // value
+  , boost::use_default
+  , Value_type // ref
```

Comment:
```suggestion
  , std::conditional_t<std::is_same<I,Value_type>::value, boost::use_default, Value_type> // ref
```

Reply:
I don't remember precisely but it has to do with boost iterator being very much picky about a particular type of iterator. Now I remember. There is an issue with boost iterator adaptor that follows the standard saying that a bidirectional iterator must have a reference that is a reference type. If not, the adapted iterator will not be directional. We ended up with an infinite loop in `std::prev` because of that.

Reply:
That is actually [*LegacyForwardIterator*](https://en.cppreference.com/w/cpp/named_req/ForwardIterator) that requires that `reference` is a reference type:
> Let T be the value type of It. The type `std::iterator_traits<It>::reference` must be either 
> 
>  - `T&` or `T&&` (since C++11) if It satisfies [LegacyOutputIterator](https://en.cppreference.com/w/cpp/named_req/OutputIterator) (It is mutable), or
>  - `const T&` or `const T&&` (since C++11) otherwise (It is constant), 
> 
> (where T is the type denoted by `std::iterator_traits<It>::value_type`)

Reply:
Then maybe the second argument of `Prevent_deref` should be the reference type `Ref_type`, directly (with a `static_assert<std::is_reference_v<Ref_type>>`).